sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupByKey.java

```diff
@@ -162,6 +164,45 @@ public static void applicableTo(PCollection<?> input) {
       throw new IllegalStateException(
           "GroupByKey must have a valid Window merge function.  " + "Invalid because: " + cause);
     }
+
+    // Validate that the trigger does not finish before garbage collection time
+    if (!triggerIsSafe(windowingStrategy)) {
```

Comment:
Is this logic GBK specific? Perhaps applies to stateful ParDo and Combine as well?

Reply:
Correction - stateful ParDo doesn't use triggers. But Combine might still be affected by this?

Reply:
Combine is not a primitive operation. If you look at Combine.java you can see that it expands to include a GroupByKey. When a runner sees it in a pipeline, it will usually replace the operation. But the validation happens on the raw graph.